### DISCUSSION TITLE
[WASM] Page Loading Fails - Potentially because 2 nested XAML controls in a direct parent/child  get instantiated/InitializeComponented/loaded in different order between WASM / WinUI

### DISCUSSION THREAD

**Discussion by ToddGlodek** [unoplatform/uno#10276] [General]
My app uses DependencyProperty.Registration to create XAML bindable properties for our custom charting controls.  On WinUI  it works fine - but not on WASM.  

I observe in the Uno documentation & code that Uno does not support PropertyMetadata.Create(...) on WASM  because there is a NotImplemented attribute on the Create() method.
```
    [NotImplemented(new string[] { "__ANDROID__", "__IOS__", "NET461", "__WASM__", "__SKIA__", "__NETSTD_REFERENCE__", "__MACOS__" })]
    public static PropertyMetadata Create(object defaultValue)
    {
        throw new NotImplementedException("The member PropertyMetadata PropertyMetadata.Create(object defaultValue) is not implemented in Uno.");
    }
```

However, there is no NotImplemented attribute on the constructor - so I would infer that it should work.
```
        public PropertyMetadata(object defaultValue)
        {
            DefaultValue = defaultValue;
        }
```

Yet, when I create an instance of one of these controls, the app simply stops executing ( yet doesn't crash).   It just wont create an instance of the class ( no log entries / console.writelines  placed inside the constructor are observed ) when using WASM.    Although, again, in WinUI it works fine.  

One thing that  ?MIGHT? be related is that  when I navigate away from the page that contains the control - and then try to return ( i.e. display the control a second time ).   I see this error in the output:
```
Unhandled Exception:
System.InvalidOperationException: Handle is not initialized.
   at System.Runtime.InteropServices.GCHandle.Free() in D:\a\Uno.DotnetRuntime.WebAssembly\Uno.DotnetRuntime.WebAssembly\runtime\src\libraries\System.Private.CoreLib\src\System\Runtime\InteropServices\GCHandle.cs:line 75
   at Microsoft.UI.Xaml.UIElement.Finalize() in C:\a\1\s\src\Uno.UI\UI\Xaml\UIElement.wasm.cs:line 99

Unhandled Exception:
System.InvalidOperationException: Handle is not initialized.
   at System.Runtime.InteropServices.GCHandle.Free() in D:\a\Uno.DotnetRuntime.WebAssembly\Uno.DotnetRuntime.WebAssembly\runtime\src\libraries\System.Private.CoreLib\src\System\Runtime\InteropServices\GCHandle.cs:line 75
   at Microsoft.UI.Xaml.UIElement.Finalize() in C:\a\1\s\src\Uno.UI\UI\Xaml\UIElement.wasm.cs:line 99
[ERROR] FATAL UNHANDLED EXCEPTION: System.InvalidOperationException: Handle is not initialized.
   at System.Runtime.InteropServices.GCHandle.Free() in D:\a\Uno.DotnetRuntime.WebAssembly\Uno.DotnetRuntime.WebAssembly\runtime\src\libraries\System.Private.CoreLib\src\System\Runtime\InteropServices\GCHandle.cs:line 75
   at Microsoft.UI.Xaml.UIElement.Finalize() in C:\a\1\s\src\Uno.UI\UI\Xaml\UIElement.wasm.cs:line 99
Uncaught ExitStatus ExitStatus
    at quit_ (localhost꞉5000/package_c44664c3271a9a33255de72c47b8bca8f6bb75a1/dotnet.js:1:251)
    at exit (localhost꞉5000/package_c44664c3271a9a33255de72c47b8bca8f6bb75a1/dotnet.js:1:239961)
    at _exit (localhost꞉5000/package_c44664c3271a9a33255de72c47b8bca8f6bb75a1/dotnet.js:1:111767)
```

I haven't tried to troubleshoot this yet, so it could be entirely unrelated.

**Comment by ToddGlodek** [comment]
UPDATE :  One thing I noticed (not sure if it is important or just an oddity of the way logging happens) ---  but the order in which the various objects is getting instantiated appears to be different between WinUI & WASM:

On WinUI - this is what happens
```
MyProjectWinUI3.MainPage: Debug: OnNavigationRequest: Navigating to : target=FlowModelDetails
MyProjectWinUI3.MainPage: Debug: NavigateToFlowModelDetails: Navigating To FlowModelDetailsPage_ForEach
MyProjectWinUI3.Shared.FlowModelDetailsPageAbstract: Debug: .ctor: _ctor FlowModelDetailsPageAbstract
MyProjectWinUI3.Shared.FlowModelDetailsPage_ForEach: Debug: .ctor: _ctor FlowModelDetailsPage_ForEach

MyProjectWinUI3.ViewModels.FlowModelDetailsPage_ForEachViewModel: Debug: .ctor: _ctor FlowModelDetailsPage_ForEachViewModel
MyProjectWinUI3.Shared.ResultsDistributionLayoutControlAbstract: Debug: .ctor: _ctor ResultsDistributionLayoutControlAbstract
MyProjectWinUI3.Shared.FlowDetailsLayoutControl: Debug: .ctor: _ctor FlowDetailsLayoutControl
```

But on WASM the order is different (observe the three entries between the blank lines).   Plus on WInUI the log continues to to capture all the expected events, but on WASM we just see all these "The thread 0x????? has exited" statements.
```
dbug: MyProjectWinUI3.MainPage[0]      OnNavigationRequest: Navigating to : target=FlowModelDetails
dbug: MyProjectWinUI3.MainPage[0]      NavigateToFlowModelDetails: Navigating To FlowModelDetailsPage_ForEach
dbug: MyProjectWinUI3.Shared.FlowModelDetailsPageAbstract[0]      .ctor: _ctor FlowModelDetailsPageAbstract
dbug: MyProjectWinUI3.Shared.FlowModelDetailsPage_ForEach[0]      .ctor: _ctor FlowModelDetailsPage_ForEach

dbug: MyProjectWinUI3.Shared.ResultsDistributionLayoutControlAbstract[0]      .ctor: _ctor ResultsDistributionLayoutControlAbstract
dbug: MyProjectWinUI3.Shared.FlowDetailsLayoutControl[0]      .ctor: _ctor FlowDetailsLayoutControl
dbug: MyProjectWinUI3.ViewModels.FlowModelDetailsPage_ForEachViewModel[0]      .ctor: _ctor FlowModelDetailsPage_ForEachViewModel

The thread 0x3bcc has exited with code 0 (0x0).
The thread 0x7054 has exited with code 0 (0x0).
The thread 0x2b4 has exited with code 0 (0x0).
The thread 0x53f0 has exited with code 0 (0x0).
The thread 0x632c has exited with code 0 (0x0).
The thread 0x3e10 has exited with code 0 (0x0).
The thread 0x7448 has exited with code 0 (0x0).
The thread 0x58cc has exited with code 0 (0x0).
The thread 0x8f8 has exited with code 0 (0x0).
The thread 0x784 has exited with code 0 (0x0).
The thread 0x61b4 has exited with code 0 (0x0).
The thread 0xd88 has exited with code 0 (0x0).
The thread 0x3b14 has exited with code 0 (0x0).
The thread 0x57f4 has exited with code 0 (0x0).
The thread 0x474c has exited with code 0 (0x0).
The thread 0x194 has exited with code 0 (0x0).
The thread 0x3ed4 has exited with code 0 (0x0).
The thread 0x47c4 has exited with code 0 (0x0).
The thread 0x54b4 has exited with code 0 (0x0).
The thread 0x36b8 has exited with code 0 (0x0).
The thread 0x3714 has exited with code 0 (0x0).
The thread 0x1f6c has exited with code 0 (0x0).
```

**Reply by jeromelaban** [reply]
Thanks for the details. If you can provide a repro for your issue, it'll be useful. 

The messages `The thread 0x3bcc has exited with code 0 (0x0).` are not from the Wasm app, but rather the Web Server process, so those are not relevant.

Note that the order of some events for the `Frame` can be different because of https://github.com/unoplatform/uno/issues/3519. If you get a repro working, we can see how to get around it.

**Comment by ToddGlodek** [comment]
Ok, I get that "The thread has exited" is part of the execution process - but on WinUI it goes on and logs all the rest of the content related to the control loading - whereas on WASM it just stops.

Also - overnight I removed all the references to DependencyProperty.Registration & commented out all the XAML references to the property.   It still just suspends - but it isn't related to  DependencyProperty as I initially thought.

**Comment by ToddGlodek** [comment]
Yes, I am defiantly using a XAML to define " [2 controls in a direct parent/child relationship (on the logical tree) called outer and inner](https://github.com/unoplatform/uno/issues/3519)"  exactly as detailed in https://github.com/unoplatform/uno/issues/3519.

I will remove the nesting, and load the controls programmatically in code to see what happens.   I'll get back to you.

**Comment by ToddGlodek** [comment]
Upon closer inspection,  I am noticing when I UNLOAD the WinUI project head and then only have the WASM project head - I wee this blue squiggly warning in the XAML that "The type Control Template does not support direct content".  I don't see this when the Windows project is loaded.   

I am assuming that this is similar to the red squiggly you see related to InitializeComponent()  and other items in the code that have been named inside XAML with the x:Name="MyControl".     This is only a VisualStudio artifact, right?   Its not a limitation of WASM vs WinUI ?

**Reply by jeromelaban** [reply]
This is indeed a Visual Studio issue. There's been progress on this in VS 17.4, which should be released at .NET conf

**Comment by ToddGlodek** [comment]
OK - I spend 3 days tying to get this to work - but its an absolutely no better.  I added a bunch of Logging - and I am finding that it is not just an issue with the order in which the constructors are called, but there are WinUI/WASM differences in the sequence in which the DataContext is set and the  Load events are called.  

It seems like on WinUI all the constructors are called - then the DataContext ( InitializeComponented() ) is set on the leaf nodes before the branches and the trunk, and then finally the load event (starting with the leaf nodes first).  

On WinUI - this is what happens
```
MyProjectWinUI3.MainPage: Debug: NavigateToFlowModelDetails: Navigating To FlowModelDetailsPage_ForEach
MyProjectWinUI3.Shared.FlowModelDetailsPageAbstract: Debug: .ctor
MyProjectWinUI3.Shared.FlowModelDetailsPage_ForEach: Debug: .ctor
MyProjectWinUI3.ViewModels.FlowModelDetailsPage_ForEachViewModel: Debug: .ctor
MyProjectWinUI3.Shared.ResultsDistributionContentControlAbstract: Debug: .ctor
MyProjectWinUI3.Shared.FlowDetailsContentControl: Debug: .ctor

MyProject.Controls.ResultsDistributionCanvas: Debug: _DataContextChanged
MyProject.Controls.ResultsDistributionCanvas: Debug: TryRedraw: No Redraw: MyProject.Controls.ResultsDistributionCanvas (IsLoaded == false)
MyProjectWinUI3.Shared.FlowDetailsContentControl: Debug: On_FlowModelChanged: Details Type is TriangularEstimatesForEachCanvas
MyProjectWinUI3.Shared.FlowDetailsContentControl: Debug: SetFlowDetailsContent: Setting Content to instatnc of detailsType
MyProject.Controls.TriangularEstimatesForEachCanvas: Debug: _DataContextChanged
MyProject.Controls.TriangularEstimatesForEachCanvas: Debug: TryRedraw: No Redraw: MyProject.Controls.TriangularEstimatesForEachCanvas (IsLoaded == false)
MyProjectWinUI3.Shared.FlowDetailsContentControl: Debug: SetFlowDetailsContent: Content Set
MyProject.Controls.DomainRangeCanvasAbstract: Debug: OnPropertyChanged: Display Bounds [4]:[100]-[0]:[50] 
MyProject.Controls.TriangularEstimatesForEachCanvas: Debug: TryRedraw: No Redraw: MyProject.Controls.TriangularEstimatesForEachCanvas (IsLoaded == false)
MyProject.Controls.TriangularEstimatesForEachCanvas: Debug: TryRedraw: No Redraw: MyProject.Controls.TriangularEstimatesForEachCanvas (IsLoaded == false)
MyProject.Controls.TriangularEstimatesForEachCanvas: Debug: TryRedraw: No Redraw: MyProject.Controls.TriangularEstimatesForEachCanvas (IsLoaded == false)
MyProjectWinUI3.Shared.FlowDetailsContentControl: Debug: On_FlowModelChanged: Details Type is TriangularEstimatesForEachCanvas
MyProjectWinUI3.Shared.FlowDetailsContentControl: Debug: SetFlowDetailsContent: Setting Content to instatnc of detailsType
MyProject.Controls.TriangularEstimatesForEachCanvas: Debug: _DataContextChanged
MyProject.Controls.TriangularEstimatesForEachCanvas: Debug: TryRedraw: No Redraw: MyProject.Controls.TriangularEstimatesForEachCanvas (IsLoaded == false)
MyProjectWinUI3.Shared.FlowDetailsContentControl: Debug: SetFlowDetailsContent: Content Set
MyProject.Controls.DomainRangeCanvasAbstract: Debug: OnPropertyChanged: Display Bounds [4]:[100]-[0]:[50] 
MyProject.Controls.TriangularEstimatesForEachCanvas: Debug: TryRedraw: No Redraw: MyProject.Controls.TriangularEstimatesForEachCanvas (IsLoaded == false)
MyProject.Controls.TriangularEstimatesForEachCanvas: Debug: TryRedraw: No Redraw: MyProject.Controls.TriangularEstimatesForEachCanvas (IsLoaded == false)
MyProject.Controls.TriangularEstimatesForEachCanvas: Debug: TryRedraw: No Redraw: MyProject.Controls.TriangularEstimatesForEachCanvas (IsLoaded == false)
MyProject.Controls.ResultsDistributionCanvas: Debug: _DataContextChanged
MyProject.Controls.ResultsDistributionCanvas: Debug: TryRedraw: No Redraw: MyProject.Controls.ResultsDistributionCanvas (IsLoaded == false)
MyProject.Controls.TriangularEstimatesForEachCanvas: Debug: _DataContextChanged
MyProject.Controls.TriangularEstimatesForEachCanvas: Debug: TryRedraw: No Redraw: MyProject.Controls.TriangularEstimatesForEachCanvas (IsLoaded == false)
MyProject.Controls.TriangularEstimatesForEachCanvas: Debug: TryRedraw: No Redraw: MyProject.Controls.TriangularEstimatesForEachCanvas (IsLoaded == false)
MyProject.Controls.ResultsDistributionCanvas: Debug: TryRedraw: No Redraw: MyProject.Controls.ResultsDistributionCanvas (IsLoaded == false)
MyProjectWinUI3.Shared.FlowDetailsContentControl: Debug: OnLoaded
MyProjectWinUI3.Shared.FlowModelDetailsPage_ForEach: Debug: FlowModelDetailsPage_ForEach_Loaded:  this.FlowDetailsContent.IsLoaded: True
MyProject.Controls.CanvasAbstract: Debug: _OnLoaded
MyProject.Controls.ResultsDistributionCanvas: Debug: TryRedraw: Drew: MyProject.Controls.ResultsDistributionCanvas
MyProject.Controls.ResultsDistributionCanvas: Debug: TryRedraw: Drew: MyProject.Controls.ResultsDistributionCanvas
MyProject.Controls.CanvasAbstract: Debug: _OnLoaded
MyProject.Controls.TriangularEstimatesForEachCanvas: Debug: TryRedraw: Drew: MyProject.Controls.TriangularEstimatesForEachCanvas
MyProject.Controls.ModelControlAbstract: Debug: InitPopup: Popup initialization Failed
MyProject.Controls.ModelControlAbstract: Debug: InitPopup: Popup initialization Failed
MyProject.Controls.ModelControlAbstract: Debug: InitPopup: Popup initialization Failed
MyProject.Controls.DomainRangeCanvasAbstract: Debug: OnPropertyChanged: Display Bounds [0]:[50]-[0]:[0.05] 
MyProject.Controls.ResultsDistributionCanvas: Debug: TryRedraw: Drew: MyProject.Controls.ResultsDistributionCanvas
```
But on WASM - not only is the order of the constructors different - the InitializeComponented() and Load() events appear to happen in a different order between WinUI & WASM.  On WASM the page level Loaded event happens first,  then the leaf nodes try to initialize.
```
dbug: MyProjectWinUI3.MainPage[0]      NavigateToFlowModelDetails: Navigating To FlowModelDetailsPage_ForEach

dbug: MyProjectWinUI3.Shared.FlowModelDetailsPageAbstract[0]      .ctor
dbug: MyProjectWinUI3.Shared.FlowModelDetailsPage_ForEach[0]      .ctor
dbug: MyProjectWinUI3.Shared.ResultsDistributionContentControlAbstract[0]
dbug: MyProjectWinUI3.Shared.FlowDetailsContentControl[0]      .ctor


dbug: MyProjectWinUI3.ViewModels.FlowModelDetailsPage_ForEachViewModel[0]           .ctor

dbug: MyProjectWinUI3.Shared.FlowModelDetailsPage_ForEach[0]      FlowModelDetailsPage_ForEach_Loaded:  this.FlowDetailsContent.IsLoaded: False
dbug: MyProjectWinUI3.Shared.FlowDetailsContentControl[0]      OnLoaded
dbug: MyProject.Controls.ResultsDistributionCanvas[0]      _DataContextChanged
dbug: MyProject.Controls.ResultsDistributionCanvas[0]      TryRedraw: No Redraw: MyProject.Controls.ResultsDistributionCanvas (IsLoaded == false)

< ABENDS >
```
 
 However, because the DataContext does not appear to have been set before the Loaded event is fired on WASM, the control's view model data does not seem to be available to continue the rendering process then it simply ABENDS after only a few steps.
 
 This issue feels identical to the issue I opened two years ago _Uno.Shared Page/UserControl Events Fired In Different Order ( UPW vs WASM )_ #4072 ( which was flagged as a duplicate of _Loading, Loaded & Unloaded events are called too early_ #2895 ).   However, it isn't clear that just changing just the _Loading, Loaded & Unloaded order fixes the entire DataContext ( InitializeComponented() ) sequence issue - maybe it will.   I also see that there is also  _[EPIC] Controls initialization sequence_ #3519 - which is more comprehensive.  
 
 At this point, given that we originally opened #4072 more than two years ago - and and are still unable to render our custom controls - I feel like I have to report to my management that Uno on WASM is not fit for purpose.

**Reply by jeromelaban** [reply]
As I mentioned previously, if you can get a specific repro to your issue (an individual project based on the Uno template, demonstrating explicitly the issue), we may be able to find a different way to get you going. It looks like you have that somehow, so can you make it available?

The important part for the repro is the scenario you're trying to get to, not necessarily the issue you're hitting.

**Comment by ToddGlodek** [comment]
Per your request, I have mocked up a [very simple repository](https://github.com/ToddGlodek/UnoBug2895Test) that demonstrates the divergence between WinUI and WASM.    This illustrates the issues in as little code as possible.

[UnoBug2895Test-main.zip](https://github.com/unoplatform/uno/files/9914781/UnoBug2895Test-main.zip)

**Reply by jeromelaban** [reply]
Thanks, this will help!

**Comment by ToddGlodek** [comment]
So I augmented the test harness app in the [UnoBug2895Test repository](https://github.com/ToddGlodek/UnoBug2895Test) (described above) such that it renders three shapes on the left canvas.  The shapes are passed as a list, from MainPage => SideBySideGridPage (via OnNavigatedTo() parameter) =>  ViewModel_LeftCanvas.  The LeftCanvas control then renders each of the shapes in response to an PropertyChanged event emitted from the view model.  Each control in the list is positioned on top previous control, and is sized to be slightly smaller than its predecessor.   The canvas reacts to OnSizeChanged() events, and resizes each of the controls appropriately, given the size of the control just beneath it.  

This is a reasonable chain of events that both WinUI & WASM should be able to support without any platform specific customizations.  However, where it works fine on WinUI, it fails on WASM.    

<img width="1118" alt="3ShapesComposite" src="https://user-images.githubusercontent.com/19764509/199653815-38b8ecca-52de-4409-9b56-de8aa9b60131.png">

On  WASM, only the first control is rendered properly (even though all three are in the canvas' Children collection).  Because each control relies on size of the previous control to determine its own size, WASM is unable to compute the height/width of the  the 2nd/3rd controls  - and simply sets their size to zero.

NOTE how on WinUI the Height/Weight ( and ActualHeight/ActualWeight ) values are properly initialized by the second resize event.
```
UnoBug2895Test.Views.SideBySideGridPage: Critical: .ctor
UnoBug2895Test.ViewModels.ShapesDisplayPageVieModel: Critical: .ctor
UnoBug2895Test.Views.CanvasAbstract: Critical: .ctor
UnoBug2895Test.ViewModels.ViewModelAbstract: Critical: .ctor
UnoBug2895Test.ViewModels.ViewModelLeftCanvas: Critical: .ctor
UnoBug2895Test.Views.CanvasAbstract: Critical: .ctor
UnoBug2895Test.ViewModels.ViewModelAbstract: Critical: .ctor
UnoBug2895Test.Views.SideBySideGridPage: Critical: OnNavigatedTo
UnoBug2895Test.Views.SideBySideGridPage: Critical: Changed: ShapesList
UnoBug2895Test.Views.SideBySideGridPage: Critical: OnDataContextChanged
UnoBug2895Test.Views.LeftCanvas: Critical: OnSizeChanged
UnoBug2895Test.Views.LeftCanvas: Critical: OnLoaded
UnoBug2895Test.Views.SideBySideGridPage: Critical: OnLoaded
UnoBug2895Test.Views.LeftCanvas: Critical: Changed: Shapes
UnoBug2895Test.Views.LeftCanvas: Critical: Resizing : Resizing : Microsoft.UI.Xaml.Shapes.Rectangle Z(1) H/W(NaN/NaN) AH/AW(0/0)
UnoBug2895Test.Views.LeftCanvas: Critical: Resizing : Resizing : Microsoft.UI.Xaml.Shapes.Ellipse Z(2) H/W(NaN/NaN) AH/AW(0/0)
UnoBug2895Test.Views.LeftCanvas: Critical: Resizing : Resizing : Microsoft.UI.Xaml.Shapes.Rectangle Z(3) H/W(NaN/NaN) AH/AW(0/0)
UnoBug2895Test.Views.LeftCanvas: Critical: Changed: Shapes
UnoBug2895Test.Views.LeftCanvas: Critical: Resizing : Resizing : Microsoft.UI.Xaml.Shapes.Rectangle Z(1) H/W(710/609) AH/AW(710/609)
UnoBug2895Test.Views.LeftCanvas: Critical: Resizing : Resizing : Microsoft.UI.Xaml.Shapes.Ellipse Z(2) H/W(568/487) AH/AW(568/487)
UnoBug2895Test.Views.LeftCanvas: Critical: Resizing : Resizing : Microsoft.UI.Xaml.Shapes.Rectangle Z(3) H/W(455/390) AH/AW(455/390)
UnoBug2895Test.Views.LeftCanvas: Critical: Changed: Shapes
UnoBug2895Test.Views.LeftCanvas: Critical: Resizing : Resizing : Microsoft.UI.Xaml.Shapes.Rectangle Z(1) H/W(710/609) AH/AW(710/609)
UnoBug2895Test.Views.LeftCanvas: Critical: Resizing : Resizing : Microsoft.UI.Xaml.Shapes.Ellipse Z(2) H/W(568/487) AH/AW(568/487)
UnoBug2895Test.Views.LeftCanvas: Critical: Resizing : Resizing : Microsoft.UI.Xaml.Shapes.Rectangle Z(3) H/W(455/390) AH/AW(455/390)
```

However on WASM,  the  Height/Weight ( and ActualHeight/ActualWeight ) values never change from ZERO
```
crit: UnoBug2895Test.Views.SideBySideGridPage[0]      .ctor
crit: UnoBug2895Test.Views.CanvasAbstract[0]      .ctor
crit: UnoBug2895Test.ViewModels.ViewModelAbstract[0]      .ctor
crit: UnoBug2895Test.ViewModels.ViewModelLeftCanvas[0]      .ctor
crit: UnoBug2895Test.Views.CanvasAbstract[0]      .ctor
crit: UnoBug2895Test.ViewModels.ViewModelAbstract[0]      .ctor
crit: UnoBug2895Test.ViewModels.ShapesDisplayPageVieModel[0]      .ctor
crit: UnoBug2895Test.Views.SideBySideGridPage[0]      OnLoaded
crit: UnoBug2895Test.Views.LeftCanvas[0]      OnLoaded
crit: UnoBug2895Test.Views.SideBySideGridPage[0]      OnNavigatedTo
crit: UnoBug2895Test.Views.SideBySideGridPage[0]      Changed: ShapesList
crit: UnoBug2895Test.Views.LeftCanvas[0]      Changed: Shapes
crit: UnoBug2895Test.Views.LeftCanvas[0]      Resizing : Resizing : Microsoft.UI.Xaml.Shapes.Rectangle Z(1) H/W(NaN/NaN) AH/AW(0/0)
crit: UnoBug2895Test.Views.LeftCanvas[0]      Resizing : Resizing : Microsoft.UI.Xaml.Shapes.Ellipse Z(2) H/W(NaN/NaN) AH/AW(0/0)
crit: UnoBug2895Test.Views.LeftCanvas[0]      Resizing : Resizing : Microsoft.UI.Xaml.Shapes.Rectangle Z(3) H/W(NaN/NaN) AH/AW(0/0)
warn: Microsoft.UI.Xaml.UIElement[0]      The BorderThicknessProperty dependency property does not exist on UnoBug2895Test.Views.LeftCanvas
crit: UnoBug2895Test.Views.LeftCanvas[0]      OnSizeChanged
crit: UnoBug2895Test.Views.LeftCanvas[0]      Resizing : Resizing : Microsoft.UI.Xaml.Shapes.Rectangle Z(1) H/W(0/0) AH/AW(0/0)
crit: UnoBug2895Test.Views.LeftCanvas[0]      Resizing : Resizing : Microsoft.UI.Xaml.Shapes.Ellipse Z(2) H/W(0/0) AH/AW(0/0)
crit: UnoBug2895Test.Views.LeftCanvas[0]      Resizing : Resizing : Microsoft.UI.Xaml.Shapes.Rectangle Z(3) H/W(0/0) AH/AW(0/0)

```

**Reply by jeromelaban** [reply]
_(When answering to a discussion, prefer answering to the thread rahter than creating a new thread. It easier to follow)_

Thanks for the updates. There are two issues here (It's generally best to have a repro for each individual one to avoid altering the behaviors). 

**The first issue** is caused the fact that Generic.xaml is not processed when defined in the main app (it's rarely used, which is why we never saw this, let's track in https://github.com/unoplatform/uno/issues/10327), causing OnApplyTemplate to not be called. To work around this, move the `generic.xaml` file to something like `Styles/Controls.xaml` and reference it from the `app.xaml` resource dictionaries:
```xaml
<Application.Resources>
    <ResourceDictionary>
        <ResourceDictionary.MergedDictionaries>
            <XamlControlsResources xmlns="using:Microsoft.UI.Xaml.Controls" />
            <ResourceDictionary Source="Styles/Controls.xaml" />
        </ResourceDictionary.MergedDictionaries>
    </ResourceDictionary>
</Application.Resources>
```

**The second issue** is related to the use of `ActualWidth` and `ActualHeight` (Similar to https://github.com/unoplatform/uno/issues/4285). Those properties generally provide values with a delay, as they are computed based on the actual layout being performed. You'll see that resizing the window makes the shapes appearm because the previous values are used for the computation. That being said, even on Windows, using these properties can cause an off-by-one use of the value, as it may use the prior value of the dimensions of the control. I suggest that you avoid using these properties, even on Windows, mainly because those are not observable.

It's likely that you'll want to create a `Panel` (with its own `MeasureOverride`/`ArrangeOverride`) instead of using of using a canvas, which can allow you to create the proper hierarchy between shapes and avoid relying on `ActualWidth`/`ActualHeight`.

**Comment by ToddGlodek** [comment]
Ok, so made the following changes in (as directed) to the bugtest repository
1.  Styles/Controls.xaml  change
2.  Converted from Canvas to Panel (implementing  MeasureOverride/ArrangeOverride). 

I was able to get that to draw the tree shapes identically on WinUI & WASM.    **However, there was still a significant inconsistency between WinUI/WASM** 
<img width="604" alt="WASM_DRAW_3Shapes" src="https://user-images.githubusercontent.com/19764509/201122658-159ffc63-872f-48ee-b495-fc408370933a.png">

WASM doesn't fire OnDataContextChanged when the DataContext is derived from a parent control in XAML hierarchy.  I was able to HACK around it but my hack wouldn't work if the parent control's DataContext instance was somehow replaced during runtime.

```

namespace UnoBug2895Test.Views;

internal class SideBySidePanel: Microsoft.UI.Xaml.Controls.Control
{

    /*************************************************************************************
     *   This demonstrates the problem with the sequence of events being fired between   *
     *   WinUI and WASM.  On WinUI the sequence is as below, but on WASM the first       *
     *   step, OnDataContextChanged() is never called - even though ViewModel is         *
     *   instantiated already BEFORE OnApplyTemplate().  (Note: there may be additional  *
     *   intermediate steps in the middle that I am not tracking)                        *
     *                                                                                   *
     *   1)  OnDataContextChanged()                                                      *
     *   2)  OnApplyTemplate()                                                           *                                                 
     *   3)  OnLoaded()                                                                  *
     *                                                                                   *
     ************************************************************************************/


    public SideBySideGrid()
    {

        this.Log().MethodInvoked();

        DataContextChanged += OnDataContextChanged;
        Loaded += OnLoaded;

        DefaultStyleKey = typeof(SideBySideGrid);


    }

    private void OnDataContextChanged(FrameworkElement sender, DataContextChangedEventArgs args)
    {
        //**** WinUI  #1
        //**** WASM @ NEVER CALLED -  Yet, the ViewModel is instantiated already BEFORE OnApplyTemplate(). 
        //****         As a consequence, no event handlers (e.g. On_PropertyChanged) that a developer would attempt 
        //****         to attach here would ever get registered.   So instead, we will attach them in OnApplyTemplate  
        //****         when HAS_UNO.
        

#if !HAS_UNO 
        var vuMod = (args.NewValue as ObservableObject);
        vuMod.PropertyChanged += On_PropertyChanged;
#endif

        this.Log().MethodInvoked();

    }
    protected override void OnApplyTemplate()
    {
        //**** WinUI #2
        //**** WASM #1

#if HAS_UNO
        var vuMod = (DataContext as ObservableObject);
        vuMod.PropertyChanged += On_PropertyChanged;
#endif

        this.Log().MethodInvoked();

        base.OnApplyTemplate();

        var booText = GetTemplateChild("txtBoo");

    }

    private void OnLoaded(object sender, Microsoft.UI.Xaml.RoutedEventArgs e)
    {
        //**** WinUI  #3
        //**** WASM #2

        this.Log().MethodInvoked();
    }

    private void On_PropertyChanged(object sender, System.ComponentModel.PropertyChangedEventArgs e)
    {
        this.Log().PropertyChanged(e);

        LeftPanel leftPanel = (LeftPanel) GetTemplateChild("LeftPanel");

        if (leftPanel.IsLoaded)
        {
            if (e.PropertyName == nameof(ShapesDisplayPageVieModel.ShapesList))
            {
                (leftPanel.DataContext as ViewModelLeftCanvas).Shapes = ((ShapesDisplayPageVieModel) sender).ShapesList;
            }
        }
    }
}

```

**Reply by jeromelaban** [reply]
Thanks for the update. If you can provide an individual repro sample for this, it will help.
